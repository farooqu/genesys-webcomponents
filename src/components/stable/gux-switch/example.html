<h1>gux-switch</h1>

<h2>Text</h2>

<h3>Small</h3>
<gux-switch
  layout="small"
  value="day"
  onchange="notify(event)"
  oninput="notify(event)"
>
  <gux-switch-item value="month">Month</gux-switch-item>
  <gux-switch-item value="week">Week</gux-switch-item>
  <gux-switch-item value="day">Day</gux-switch-item>
  <gux-switch-item value="hour" disabled>Hour</gux-switch-item>
  <gux-switch-item value="minute">Minute</gux-switch-item>
</gux-switch>

<h3>Default</h3>
<gux-switch
  layout="default"
  value="month"
  onchange="notify(event)"
  oninput="notify(event)"
>
  <gux-switch-item value="month">Month</gux-switch-item>
  <gux-switch-item value="week" disabled>Week</gux-switch-item>
  <gux-switch-item value="day">Day</gux-switch-item>
  <gux-switch-item value="hour">Hour</gux-switch-item>
  <gux-switch-item value="minute">Minute</gux-switch-item>
</gux-switch>

<h2>Icon</h2>

<h3>Small</h3>
<gux-switch
  layout="small"
  value="line"
  onchange="notify(event)"
  oninput="notify(event)"
>
  <gux-switch-item value="bar-vertical">
    <gux-icon
      icon-name="graph-bar-v"
      screenreader-text="View as virtical bar graph"
    ></gux-icon>
  </gux-switch-item>
  <gux-switch-item value="bar">
    <gux-icon
      icon-name="graph-bar-h"
      screenreader-text="View as horizontal bar graph"
    ></gux-icon>
  </gux-switch-item>
  <gux-switch-item value="grid-cell">
    <gux-icon
      icon-name="view-stacked"
      screenreader-text="View as table"
    ></gux-icon>
  </gux-switch-item>
  <gux-switch-item value="grid">
    <gux-icon icon-name="view-all" screenreader-text="View as grid"></gux-icon>
  </gux-switch-item>
  <gux-switch-item value="line">
    <gux-icon
      icon-name="graph-line"
      screenreader-text="View as line graph"
    ></gux-icon>
  </gux-switch-item>
  <gux-switch-item value="stack" disabled>
    <gux-icon
      icon-name="graph-bar-h"
      screenreader-text="View as stacked bar graph"
    ></gux-icon>
  </gux-switch-item>
</gux-switch>

<h3>Default</h3>
<gux-switch
  layout="default"
  value="bar"
  onchange="notify(event)"
  oninput="notify(event)"
>
  <gux-switch-item value="bar-vertical">
    <gux-icon
      icon-name="graph-bar-v"
      screenreader-text="View as virtical bar graph"
    ></gux-icon>
  </gux-switch-item>
  <gux-switch-item value="bar">
    <gux-icon
      icon-name="graph-bar-h"
      screenreader-text="View as horizontal bar graph"
    ></gux-icon>
  </gux-switch-item>
  <gux-switch-item value="grid-cell">
    <gux-icon
      icon-name="view-stacked"
      screenreader-text="View as table"
    ></gux-icon>
  </gux-switch-item>
  <gux-switch-item value="grid">
    <gux-icon icon-name="view-all" screenreader-text="View as grid"></gux-icon>
  </gux-switch-item>
  <gux-switch-item value="line">
    <gux-icon
      icon-name="graph-line"
      screenreader-text="View as line graph"
    ></gux-icon>
  </gux-switch-item>
  <gux-switch-item value="stack" disabled>
    <gux-icon
      icon-name="graph-bar-h"
      screenreader-text="View as stacked bar graph"
    ></gux-icon>
  </gux-switch-item>
</gux-switch>

<h2>Get value</h2>

<gux-switch id="get-value" layout="small" value="month">
  <gux-switch-item value="month">Month</gux-switch-item>
  <gux-switch-item value="week">Week</gux-switch-item>
  <gux-switch-item value="day">Day</gux-switch-item>
  <gux-switch-item value="hour">Hour</gux-switch-item>
  <gux-switch-item value="minute">Minute</gux-switch-item>
</gux-switch>

<div>
  Last retrieved value from gux-switch: <span id="value-output">N/A</span>
</div>

<gux-button
  onclick="(function() {
  const guxSwitch = document.getElementById('get-value');
  const valueOutput = document.getElementById('value-output');

  valueOutput.textContent = guxSwitch.value;
})()"
  >Get Value</gux-button
>
