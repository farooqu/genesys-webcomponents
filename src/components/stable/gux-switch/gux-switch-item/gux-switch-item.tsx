import { Component, h, Host, Listen, JSX, Prop } from '@stencil/core';

@Component({
  styleUrl: 'gux-switch-item.less',
  tag: 'gux-switch-item'
})
export class GuxSwitchItem {
  @Prop()
  value: string;

  @Prop()
  selected: boolean = false;

  @Prop()
  disabled: boolean = false;

  @Listen('click')
  onClick(e: MouseEvent): void {
    if (this.disabled) {
      e.stopPropagation();
    }
  }

  render(): JSX.Element {
    return (
      <Host class={{ 'gux-selected': this.selected }}>
        <button type="button" class="gux-switch-item" disabled={this.disabled}>
          <slot />
        </button>
      </Host>
    ) as JSX.Element;
  }
}
