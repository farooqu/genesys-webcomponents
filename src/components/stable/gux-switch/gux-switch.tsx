import {
  Component,
  Element,
  h,
  Host,
  Listen,
  JSX,
  Prop,
  State
} from '@stencil/core';

import { GuxSwitchAllowedLayouts } from './gux-switch.types';

import simulateNativeEvent from '../../../utils/dom/simulate-native-event';
import { trackComponent } from '../../../usage-tracking';

/**
 * @slot - list of gux-switch-item elements
 */
@Component({
  styleUrl: 'gux-switch.less',
  tag: 'gux-switch',
  shadow: true
})
export class GuxSwitch {
  @Element()
  root: HTMLElement;

  @Prop({ mutable: true })
  value: string;

  @Prop()
  layout: GuxSwitchAllowedLayouts = 'default';

  @State()
  switchItems: HTMLGuxSwitchItemElement[] = [];

  @Listen('click')
  onClick(e: MouseEvent): void {
    e.stopPropagation();

    const switchItem = (e.target as HTMLElement).closest('gux-switch-item');

    if (switchItem && this.value !== switchItem.value) {
      this.value = switchItem.value;

      simulateNativeEvent(this.root, 'input');
      simulateNativeEvent(this.root, 'change');
    }
  }

  private slotChanged(): void {
    this.switchItems = Array.from(
      this.root.children
    ) as HTMLGuxSwitchItemElement[];
  }

  componentWillLoad(): void {
    trackComponent(this.root, { variant: this.layout });
  }

  componentWillRender(): void {
    this.switchItems.forEach(switchItem => {
      switchItem.selected = switchItem.value === this.value;
    });
  }

  render(): JSX.Element {
    return (
      <Host role="group" class={`gux-${this.layout}`}>
        <slot onSlotchange={this.slotChanged.bind(this)} />
      </Host>
    ) as JSX.Element;
  }
}
