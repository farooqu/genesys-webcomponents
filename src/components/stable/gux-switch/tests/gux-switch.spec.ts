import { newSpecPage } from '@stencil/core/testing';

import { GuxSwitch } from '../gux-switch';
import { GuxSwitchItem } from '../gux-switch-item/gux-switch-item';

const components = [GuxSwitch, GuxSwitchItem];
const html = `
  <gux-switch layout="small" value="day">
    <gux-switch-item value="month">Month</gux-switch-item>
    <gux-switch-item value="week">Week</gux-switch-item>
    <gux-switch-item value="day">Day</gux-switch-item>
    <gux-switch-item value="hour" disabled>Hour</gux-switch-item>
    <gux-switch-item value="minute">Minute</gux-switch-item>
  </gux-switch>
`;
const language = 'en';

describe('gux-switch', () => {
  beforeAll(async () => {
    (
      global as NodeJS.Global & {
        InputEvent: typeof Event;
      }
    ).InputEvent = Event;
  });

  describe('#render', () => {
    it(`should render as expected`, async () => {
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxSwitch);
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('#interactions', () => {
    it(`should change value on gux-switch-item click`, async () => {
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxSwitchElement;
      const guxSwitchItemMinute: HTMLGuxSwitchItemElement =
        page.root.querySelector('gux-switch-item[value=minute]');

      expect(element.value).toBe('day');

      guxSwitchItemMinute.click();
      await page.waitForChanges();

      expect(element.value).toBe(guxSwitchItemMinute.value);
    });

    it(`should not change value on gux-switch-item click if it is disabled`, async () => {
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxSwitchElement;
      const guxSwitchItemHour: HTMLGuxSwitchItemElement =
        page.root.querySelector('gux-switch-item[value=hour]');
      const currentValue = element.value;

      expect(currentValue).toBe('day');

      guxSwitchItemHour.click();
      await page.waitForChanges();

      expect(element.value).toBe(currentValue);
    });

    it(`should emit a 'change' and 'input' event when a new item is selected`, async () => {
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxSwitchElement;
      const guxSwitchItemMinute: HTMLGuxSwitchItemElement =
        page.root.querySelector('gux-switch-item[value=minute]');
      const changeEventSpy = jest.fn();
      const inputEventSpy = jest.fn();

      element.addEventListener('change', () => {
        changeEventSpy();
      });
      element.addEventListener('input', () => {
        inputEventSpy();
      });

      guxSwitchItemMinute.click();
      await page.waitForChanges();

      expect(changeEventSpy).toHaveBeenCalledWith();
      expect(inputEventSpy).toHaveBeenCalledWith();
    });

    it(`should not emit a 'change' and 'input' event when the selected item is reselected`, async () => {
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxSwitchElement;
      const guxSwitchItemDay: HTMLGuxSwitchItemElement =
        page.root.querySelector('gux-switch-item[value=day]');
      const changeEventSpy = jest.fn();
      const inputEventSpy = jest.fn();

      element.addEventListener('change', () => {
        changeEventSpy();
      });
      element.addEventListener('input', () => {
        inputEventSpy();
      });

      guxSwitchItemDay.click();
      await page.waitForChanges();

      expect(changeEventSpy).not.toHaveBeenCalled();
      expect(inputEventSpy).not.toHaveBeenCalled();
    });

    it(`should not emit a 'change' or 'input' event when a disabled item is selected`, async () => {
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLGuxSwitchElement;
      const guxSwitchItemHour: HTMLGuxSwitchItemElement =
        page.root.querySelector('gux-switch-item[value=hour]');
      const changeEventSpy = jest.fn();
      const inputEventSpy = jest.fn();

      element.addEventListener('change', () => {
        changeEventSpy();
      });
      element.addEventListener('input', () => {
        inputEventSpy();
      });

      guxSwitchItemHour.click();
      await page.waitForChanges();

      expect(changeEventSpy).not.toHaveBeenCalled();
      expect(inputEventSpy).not.toHaveBeenCalled();
    });
  });

  describe('onSlotchange', () => {
    it(`should set selected items as expected`, async () => {
      const page = await newSpecPage({ components, html, language });
      const element = document.querySelector('gux-switch');

      const newItem = document.createElement('gux-switch-item');
      newItem.appendChild(document.createTextNode('Second'));
      newItem.setAttribute('value', 'second');
      element.appendChild(newItem);

      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();
    });

    // onSlotchange is not getting called as expected in the unit tests so it tested manually below
    it('manually call onSlotchange methods', async () => {
      const page = await newSpecPage({ components, html, language });
      const elementController: GuxSwitch = page.rootInstance;

      expect(elementController.switchItems).toHaveLength(0);

      (elementController as any).slotChanged();
      elementController.componentWillRender();

      expect(elementController.switchItems).toHaveLength(5);
    });
  });
});
