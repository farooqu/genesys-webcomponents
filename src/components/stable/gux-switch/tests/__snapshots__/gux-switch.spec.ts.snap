// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-switch #render should render as expected 1`] = `
<gux-switch class="gux-small" layout="small" role="group" value="day">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <gux-switch-item value="month">
    <!---->
    <button class="gux-switch-item" type="button">
      Month
    </button>
  </gux-switch-item>
  <gux-switch-item value="week">
    <!---->
    <button class="gux-switch-item" type="button">
      Week
    </button>
  </gux-switch-item>
  <gux-switch-item value="day">
    <!---->
    <button class="gux-switch-item" type="button">
      Day
    </button>
  </gux-switch-item>
  <gux-switch-item disabled="" value="hour">
    <!---->
    <button class="gux-switch-item" disabled="" type="button">
      Hour
    </button>
  </gux-switch-item>
  <gux-switch-item value="minute">
    <!---->
    <button class="gux-switch-item" type="button">
      Minute
    </button>
  </gux-switch-item>
</gux-switch>
`;

exports[`gux-switch onSlotchange should set selected items as expected 1`] = `
<gux-switch class="gux-small" layout="small" role="group" value="day">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <gux-switch-item value="month">
    <!---->
    <button class="gux-switch-item" type="button">
      Month
    </button>
  </gux-switch-item>
  <gux-switch-item value="week">
    <!---->
    <button class="gux-switch-item" type="button">
      Week
    </button>
  </gux-switch-item>
  <gux-switch-item value="day">
    <!---->
    <button class="gux-switch-item" type="button">
      Day
    </button>
  </gux-switch-item>
  <gux-switch-item disabled="" value="hour">
    <!---->
    <button class="gux-switch-item" disabled="" type="button">
      Hour
    </button>
  </gux-switch-item>
  <gux-switch-item value="minute">
    <!---->
    <button class="gux-switch-item" type="button">
      Minute
    </button>
  </gux-switch-item>
  <gux-switch-item value="second">
    <!---->
    <button class="gux-switch-item" type="button">
      Second
    </button>
  </gux-switch-item>
</gux-switch>
`;
